suitcase/utils/__init__.py

```diff
@@ -1,25 +1,243 @@
+        self._reserved_names = set()
+        self._artifacts = collections.defaultdict(list)
+        self.buffers = {}  # maps postfixes to buffer objects
+
+    @property
+    def artifacts(self):
+        return dict(self. _artifacts)
+
+    def reserve_name(self, label, postfix):
+        """
+        Ask the wrapper for a filepath.
+
+        An external library that needs a filepath (not a handle)
+        may use this instead of the ``open`` method.
+
+        Parameters
+        -------
+        label : string
+            partial file name (i.e. stream name)
+        postfix : string
+            relative file path and filename
+
+        Returns
+        ----------
+        filepath : Path
+         """
+        raise SuitcaseUtilsTypeError(
+            "MemoryBuffersWrapper is incompatible with exporters that require "
+            "explicit filenames.")
+
+    def open(self, label, postfix, mode, encoding=None, errors=None):
+        """
+        Request a file handle.
+
+        Like the built-in open function, this may be used as a context manager.
+
+        Parameters
+        -------
+        label : string
+            partial file name (i.e. stream name)
+        postfix : string
+            relative file path and filename
+        mode : {'x', 'xt', xb'}
+            'x' or 'xt' for text, 'xb' for binary
+        encoding : string or None
+            Not used. Accepted for compatibility with built-in open().
+        errors : string or None
+            Not used. Accepted for compatibility with built-in open().
+
+       Returns
+       -------
+       file : handle
+        """
+        # Of course, in-memory buffers have no filepath, but we still expect
+        # postfix to be a thing that looks like a relative filepath, and we use
+        # it as a unique identifier for a given buffer.
+        if Path(postfix).is_absolute():
+            raise SuitcaseUtilsValueError(
+                f"{postfix} must be structured like a relative file path.")
+        name = Path(postfix).expanduser().resolve()
+        if name in self._reserved_names:
+            raise SuitcaseUtilsValueError(
+                f"The postfix {postfix!r} has already been used.")
+        self._reserved_names.add(name)
+        self._artifacts[label].append(name)
+        if mode in ('x', 'xt'):
+            buffer = PersistentStringIO()
+        elif mode == 'xb':
+            buffer = PersistentBytesIO()
+        else:
+            raise ModeError(
+                f'the mode passed to MemoryBuffersWrapper.open is {mode} but '
+                'needs to be one of "x", "xt" or "xb"')
+        self.buffers[postfix] = buffer
+        return buffer
+
+    def close(self):
+        '''close all buffers open by the manager
+
+        Note: this is a no-op method added for API consistency only, for the
+        details on why it is a no-op see Persistent*IO definitions in
+        `suitcase.utils.__init__`
+        '''
+        ...
```

Comment:
If you take a second look at @tacaswell's follow-up comment, you'll see he is suggesting that this should call `close()` on all its buffers, even though that will have no effect. I hadn't thought of this in my original suggestion, but I agree; if later `Persistent*IO` is changed or more likely swapped out for something else by another develop with a use case that has not occurred to us, _two_ layers of no-op may confuse them.

Reply:
fix made